internal/locate/region_cache.go

```diff
@@ -734,6 +745,48 @@ func (l *KeyLocation) String() string {
 	return fmt.Sprintf("region %s,startKey:%s,endKey:%s", l.Region.String(), kv.StrKey(l.StartKey), kv.StrKey(l.EndKey))
 }
 
+// GetBucketVersion gets the bucket version of the region.
+// If the region doesn't contain buckets, returns 0.
+func (l *KeyLocation) GetBucketVersion() uint64 {
+	if l.Buckets == nil {
+		return 0
+	}
+	return l.Buckets.GetVersion()
+}
+
+// LocateBucket returns the bucket the key is located.
+func (l *KeyLocation) LocateBucket(key []byte) *Bucket {
```

Comment:
Maybe it can be simplified like:
```go
keys := l.Buckets.GetKeys()
searchLen := len(keys) - 1
i := sort.Search(searchLen, func(i int) bool {
	return bytes.Compare(key, keys[i]) < 0
})

// buckets contains region's start/end key, so i==0 means it can't find a suitable bucket
// which can happen if the bucket information is stale.
if i == 0 ||
	// the key isn't located in the last range.
	(i == searchLen && len(keys[searchLen]) != 0 && bytes.Compare(key, keys[searchLen]) >= 0) {
	return nil
}
return &Bucket{keys[i-1], keys[i]}
```